lib/ecto/exceptions.ex

```diff
@@ -286,7 +286,17 @@ defmodule Ecto.ConstraintError do
           "The changeset has not defined any constraint."
         constraints ->
           "The changeset defined the following constraints:\n\n" <>
-            Enum.map_join(constraints, "\n", &"    * #{&1.constraint} (#{&1.type}_constraint)")
+            Enum.map_join(constraints, "\n", fn c ->
+              name =
+                case c.constraint do
+                  name when is_binary(name) ->
+                    name
+
+                  %Regex{} = r ->
+                    inspect(r)
+                end
+              "    * #{name} (#{c.type}_constraint)"
+            end)
```

Comment:
Regex doesn't implement `String.Chars`, so when storing the constraint as a String | Regex, we need to do this
I was torn between using `inspect(r)` or `Regex.source(r)`